src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -95,12 +169,168 @@ public void Dispose()
+                        break;
+                    case BinaryLogRecordKind.UninitializedPropertyRead:
+                        result = ReadUninitializedPropertyReadEventArgs();
+                        break;
+                    case BinaryLogRecordKind.PropertyInitialValueSet:
+                        result = ReadPropertyInitialValueSetEventArgs();
+                        break;
+                    case BinaryLogRecordKind.AssemblyLoad:
+                        result = ReadAssemblyLoadEventArgs();
+                        break;
+                    default:
+                        OnRecoverableReadError?.Invoke(
+                            $"BuildEvent record number {recordNumber} (serialized size: {serializedEventLength}) is of unsupported type: {recordKind}.{(SkipUnknownEvents ? " Skipping it." : string.Empty)}");
+                        if (SkipUnknownEvents && serializedEventLength > 0)
+                        {
+                            SkipBytes(serializedEventLength);
+                        }
+                        else
+                        {
+                            return null;
+                        }
+                        break;
+                }
+
+                if (fileFormatVersion >= 18)
+                {
+                    long postEventPosition = _canSeek ? binaryReader.BaseStream.Position : serializedEventLength;
+                    int bytesRead = (int)(postEventPosition - preEventPosition);
+                    if (bytesRead != serializedEventLength)
+                    {
+                        string error =
+                            $"BuildEvent record number {recordNumber} was expected to read exactly {serializedEventLength} bytes from the stream, but read {bytesRead} instead.";
+
+                        if (SkipUnknownEventParts && bytesRead < serializedEventLength)
+                        {
+                            OnRecoverableReadError?.Invoke(error);
+                            SkipBytes(serializedEventLength - bytesRead);
+                        }
+                        else
+                        {
+                            throw new InvalidDataException(error);
+                        }
+                    }
+                }
+
+                recordNumber += 1;
+            }
+
+            return result;
+        }
+
+        private void SkipBytes(int count)
+        {
+            if (_canSeek)
+            {
+                binaryReader.BaseStream.Seek(count, SeekOrigin.Current);
+            }
+            else
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(count);
+                using var _ = new CleanupScope(() => ArrayPool<byte>.Shared.Return(buffer));
+                binaryReader.BaseStream.ReadAtLeast(buffer, 0, count, throwOnEndOfStream: true);
+            }
+        }
+
+        private BinaryLogRecordKind ReadTillNextEvent(Func<BinaryLogRecordKind, bool> isPreprocessRecord)
```

Comment:
NIT: please consider more precise naming of intent "Fast forward meta records until 1st BuildEventArgs"

Reply:
`PreprocessRecordsTillNextEvent`